### PR TITLE
internal/contour: decouple the cache handler

### PR DESCRIPTION
**Pull request by jpeach**
The `contour.CacheHandler` struct contains the xDS resource caches as
embedded members. If we extract a DAG observer interface, however, we
can remove knowledge of specific resources from `CacheHandler`.
All it really needs is a set of objects that can vist the DAG when it
changes to generate any relevant xDS resources, and that can be called
on to publish those into the xDS GRPC interface.

Separating out the new `contour.ResourceCache` interface gives us the
opportunity to add new DAG consumers by only having to modify code
when the Contour server starts up. There's no longer any cache handler
plumbing needed.

Once we can compose the xDS caches under the `Observer` interface, the
need for a separate cache handler struct goes away, and we can attach
everything we need directly to the event handler.

This updates #2775.

### CODE REVIEW

**Comment by jpeach** [comment]
@stevesloka This PR came out of trying to see how we can better integrate EDS with knowledge of the DAG. It heavily conflicts with your go-control-plane refactor, so we probably want to see whether it still makes sense in conjunction with that.

**Comment by jpeach** [comment]
> @stevesloka This PR came out of trying to see how we can better integrate EDS with knowledge of the DAG. It heavily conflicts with your go-control-plane refactor, so we probably want to see whether it still makes sense in conjunction with that.

Alternatively, it might make the go-control-plane change easier 🤷

**Comment by stevesloka** [comment]
I ripped out the entire cachehandler in #2604.

**Comment by jpeach** [comment]
> I ripped out the entire cachehandler in #2604.

🥳  I was able to remove it from this PR too. It's replaced by composing DAG observers.

File: internal/contour/cachehandler.go [hunk 1]
```diff
@@ -17,55 +17,32 @@
...
-	ch.SecretCache.Update(secrets)
-}
-
-func (ch *CacheHandler) updateListeners(root dag.Visitable) {
-	listeners := visitListeners(root, &ch.ListenerVisitorConfig)
-	ch.ListenerCache.Update(listeners)
+// ResourceCache is a store of an xDS resource type. It is able to
+// visit the dag.DAG to update the its resource collection, then
+// serve those resources over xDS.
+type ResourceCache interface {
```
[the remark this answers is not in the document]
Comment (jpeach) [reply internal/contour/cachehandler.go:27]: WDYT if we pull these out into a `./internal/resources` package? That would reduce access to event handler internals; might put us in a slightly better place to make the switch to the go-control-plane snapshot cache.

File: internal/contour/handler.go [hunk 2]
```diff
@@ -24,18 +24,20 @@ import (
...
 // EventHandler implements cache.ResourceEventHandler, filters k8s events towards
-// a dag.Builder and calls through to the CacheHandler to notify it that a new DAG
+// a dag.Builder and calls through to the Observer to notify it that a new DAG
 // is available.
 type EventHandler struct {
-	dag.Builder
-
-	*CacheHandler
+	Builder  dag.Builder
+	Observer dag.Observer
```
[the remark this answers is not in the document]
Comment (jpeach) [reply internal/contour/handler.go:39]: This encourages the use of hierarchies of observers to hoist functionality out of the event handler. For example, we could decouple the DAG processing metrics with an observer that handles the metrics. One of the things I also had in mind was that with xDS snapshots, we might end up in a place where the DAG is just observed by a single snapshot, so a slice would not be needed.

File: internal/contour/cachehandler.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpeach) [reply internal/contour/cachehandler.go:27]: I renamed to `internal/contour/resources.go`

File: internal/contour/handler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpeach) [reply internal/contour/handler.go:37]: I moved metrics into an observer, so the event handler doesn't need to deal with metrics any more.

**Comment by jpeach** [comment]
@skriss @youngnick This is ready for the next review pass. Thanks :)

**Comment by jpeach** [comment]
> LGTM. I don't think we've used interface assertions much before. It feels a little bit of a weird syntax for me, but I can see that it's useful. If we're going to add them, we should either do it for everything or not do it though, I think.

I removed these.

**Pull request merged**